From 1976f9a8cb132dc48dc3e1258a493652ae2a0d9b Mon Sep 17 00:00:00 2001
From: Arulselvan M <arulselvan.m@intel.com>
Date: Tue, 21 Jul 2015 04:51:11 +0530
Subject: [PATCH] ASoC: Enhance the deferred probe for dfw_sst.bin

Enhancement to download dfw_sst.bin to be in generic path
than in the specific path.
DFW bin contains vendor specific changes kcontrols, widgets,
simple coefficients and DAPM graphs alongside their firmware text.

Change-Id: I9227fb898d11aad69b989daf83b5ca9a9b80bbf2
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-6737
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 sound/soc/intel/pcm.c | 28 +++++++++++++++++++++-------
 1 file changed, 21 insertions(+), 7 deletions(-)

diff --git a/sound/soc/intel/pcm.c b/sound/soc/intel/pcm.c
index 0724d3d232f7..c0549dd9b297 100644
--- a/sound/soc/intel/pcm.c
+++ b/sound/soc/intel/pcm.c
@@ -31,6 +31,7 @@
 #include <sound/pcm.h>
 #include <sound/pcm_params.h>
 #include <sound/soc.h>
+#include <linux/firmware.h>
 #include <sound/intel_sst_ioctl.h>
 #include <asm/platform_sst_audio.h>
 #include <asm/platform_cht_audio.h>
@@ -47,6 +48,8 @@ extern struct snd_effect_ops effects_ops;
 
 /* module parameters */
 static int dpcm_enable = 1;
+static int dfw_request;
+static int defer_probe = 1; /* defer probe is enabled */
 
 /* dpcm_enable should be =0 for mofd_v0 and =1 for mofd_v1 */
 module_param(dpcm_enable, int, 0644);
@@ -1044,21 +1047,32 @@ static const struct snd_soc_component_driver pcm_component = {
 	.name           = "pcm",
 };
 
+void dfw_fw_load_cb(const struct firmware *fw, void *context)
+{
+	if (fw == NULL) {
+		pr_err("request dfw fw failed\n");
+		return;
+	}
+
+	defer_probe = 0;
+}
 static int sst_platform_probe(struct platform_device *pdev)
 {
 	struct sst_data *sst;
 	int ret;
 	struct sst_platform_data *pdata = pdev->dev.platform_data;
-	struct file *file;
 
 	pr_debug("sst_platform_probe called\n");
+	if (pdata->dfw_enable) {
+		if (!dfw_request) {
+			request_firmware_nowait(THIS_MODULE, 1, "dfw_sst.bin",
+				&pdev->dev, GFP_KERNEL, NULL, dfw_fw_load_cb);
+			dfw_request = 1;
+		}
 
-	if(pdata->dfw_enable) {
-		file = filp_open("/etc/firmware/dfw_sst.bin", O_RDONLY, 0);
-
-		if (IS_ERR(file)) {
-			pr_info("sst_platform_probe is deferred\n");
-			return -EPROBE_DEFER;
+		if (defer_probe) {
+			pr_info("sst_platform_probe deferred\n");
+			return  -EPROBE_DEFER;
 		}
 	}
 
-- 
1.9.1

